### PR TITLE
Cb 3536 add settings panel

### PR DESCRIPTION
**Pull request by tfokina**

### CODE REVIEW

File: webapp/packages/core-plugin/src/PluginManagerService.ts [hunk 1]
```diff
@@ -12,14 +12,33 @@ import { PluginSettings } from './PluginSettings';
...
+      return coreSettings;
+    } else {
+      const coreSettings = new PluginSettings(this.productManagerService.settings, 'core.' + scope, defaults);
+      this.store.set('core.' + scope, coreSettings);
 
-  getCoreSettings<T>(scope: string, defaults: T) {
-    return new PluginSettings(this.productManagerService.settings, 'core.' + scope, defaults);
+      return coreSettings;
+    }
   }
```
Comment (Wroud) [inline webapp/packages/core-plugin/src/PluginManagerService.ts]: please return `PluginSettings | undefined` don't try to initialize the setting "on the fly"

Comment (Wroud) [reply webapp/packages/core-plugin/src/PluginManagerService.ts]: create a separate function for settings creation

File: webapp/packages/core-settings/package.json [hunk 2]
```diff
@@ -17,9 +17,10 @@
   },
   "dependencies": {
     "@cloudbeaver/core-di": "~0.1.0",
-    "@cloudbeaver/core-browser": "~0.1.0"
+    "@cloudbeaver/core-browser": "~0.1.0",
+    "@cloudbeaver/plugin-settings-panel": "~0.1.0"
```
Comment (Wroud) [inline webapp/packages/core-settings/package.json]: core packages can't depends on `plugin`

File: webapp/packages/core-settings/src/SettingsManagerService.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { injectable } from '@cloudbeaver/core-di';
+import type { FormFieldType, SettingsGroup } from '@cloudbeaver/plugin-settings-panel';
```
Comment (Wroud) [inline webapp/packages/core-settings/src/SettingsManagerService.ts]: move this types to `core-settings`

File: webapp/packages/core-settings/src/SettingsManagerService.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    this.groups.set(group.id, group);
+  }
+
+  deleteSettings(settings: ScopeSettingsItem | string) {
+    if (typeof settings === 'string') {
+      this.settings = this.settings.filter(item => item.scope !== settings);
+    } else {
+      this.settings = this.settings.filter(item => item.scope !== settings.scope);
+    }
+  }
```
Comment (Wroud) [inline webapp/packages/core-settings/src/SettingsManagerService.ts]: the method has unclear behavior (scope is not a unique identifier, you can't pass ScopeSettingsItem to delete all settings in the scope)

File: webapp/packages/plugin-data-viewer/package.json [hunk 5]
```diff
@@ -21,6 +21,7 @@
     "@cloudbeaver/plugin-navigation-tabs": "~0.1.0",
     "@cloudbeaver/plugin-navigation-tree": "~0.1.0",
     "@cloudbeaver/plugin-object-viewer": "~0.1.0",
+    "@cloudbeaver/plugin-settings-panel": "~0.1.0",
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/package.json]: remove (you probably can import all that you need from `core-settings`)

File: webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts [hunk 6]
```diff
@@ -7,6 +7,8 @@
  */
 import { injectable } from '@cloudbeaver/core-di';
 import { PluginManagerService, PluginSettings } from '@cloudbeaver/core-plugin';
+import { SettingsManagerService } from '@cloudbeaver/core-settings';
+import { createSettingsGroup, FormFieldType } from '@cloudbeaver/plugin-settings-panel';
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts]: move `createSettingsGroup` to `core-settings`

File: webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts [hunk 7]
```diff
@@ -23,9 +25,43 @@ export class DataViewerSettingsService {
   /** @deprecated Use settings instead, will be removed in 23.0.0 */
   readonly deprecatedSettings: PluginSettings<typeof defaultSettings>;
 
-  constructor(private readonly pluginManagerService: PluginManagerService) {
+  constructor(private readonly pluginManagerService: PluginManagerService, private readonly settingsManagerService: SettingsManagerService) {
     this.settings = this.pluginManagerService.getPluginSettings('data-viewer', defaultSettings);
     this.deprecatedSettings = this.pluginManagerService.getDeprecatedPluginSettings('core.app.dataViewer', defaultSettings);
+
+    const DATA_EDITOR_SETTINGS_GROUP = createSettingsGroup('Data Editor');
```
Comment (Wroud) [inline webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts]: move to a separate file and export from it (and export from plugin itself)

File: webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                .filter(settingsItem => settingsItem.groupId === groupId)
+                .map(settingsItem => ({
+                  ...settingsItem,
+                  value: getValue(settingsItem.key, settingsItem.scopeType, settingsItem.scope),
+                }))}
+              readOnly
+            />
+          </>
+        ))}
+      </Group>
```
Comment (Wroud) [inline webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx]: please create `SettingsGroup` that will accept `group and render settings for it`

File: webapp/packages/plugin-settings-panel/src/SettingsPanelPluginBootstrap.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            id: 'settingsPanel',
+            label: 'plugin_settings_panel',
+            tooltip: 'plugin_settings_panel',
+          },
+          {
+            onSelect: () => this.optionsPanelService.open(() => SettingsPanelForm),
+          },
+        ),
+      ],
+    });
```
Comment (Wroud) [inline webapp/packages/plugin-settings-panel/src/SettingsPanelPluginBootstrap.ts]: please create action ACTION_OPEN_APP_SETTINGS` (we will use it later to open settings from key combinations or from fast actions)

**Review by Wroud — CHANGES_REQUESTED** [review]

File: webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+import { SettingsInfoForm } from './SettingsInfoForm';
+
+const styles = css`
+  ColoredContainer {
+    height: 100%;
+  }
+`;
+
+export const SettingsPanelForm: React.FC = observer(function SettingsPanelForm() {
```
Comment (devnaumov) [inline webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx]: React.FC could be omitted

File: webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  const settingsManagerService = useService(SettingsManagerService);
+  const pluginManagerService = useService(PluginManagerService);
+  const style = useStyles(BASE_CONTAINERS_STYLES, styles);
+
+  const groups = Array.from(settingsManagerService.groups);
+  const settings = settingsManagerService.settings;
+
+  function getValue(key: string, scopeType: string, scope?: string) {
```
Comment (devnaumov) [inline webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx]: Seems like the "scope" must be always provided?

**Review by devnaumov — CHANGES_REQUESTED** [review]

File: webapp/packages/core-plugin/src/PluginManagerService.ts (hunk 1 above)
Comment (tfokina) [reply webapp/packages/core-plugin/src/PluginManagerService.ts]: Fixed this. Now we have a function for creating settings and a function getting settings.

File: webapp/packages/core-settings/package.json (hunk 2 above)
Comment (tfokina) [reply webapp/packages/core-settings/package.json]: Fixed this

File: webapp/packages/core-settings/src/SettingsManagerService.ts (hunk 3 above)
Comment (tfokina) [reply webapp/packages/core-settings/src/SettingsManagerService.ts]: Moved the types to the `core-settings`

File: webapp/packages/core-settings/src/SettingsManagerService.ts (hunk 4 above)
Comment (tfokina) [reply webapp/packages/core-settings/src/SettingsManagerService.ts]: I agree, fixed this

File: webapp/packages/plugin-data-viewer/package.json (hunk 5 above)
Comment (tfokina) [reply webapp/packages/plugin-data-viewer/package.json]: Fixed this

File: webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts (hunk 6 above)
Comment (tfokina) [reply webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts]: Fixed this

File: webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts (hunk 7 above)
Comment (tfokina) [reply webapp/packages/plugin-data-viewer/src/DataViewerSettingsService.ts]: Fixed this + also for other settings services

File: webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx (hunk 8 above)
Comment (tfokina) [reply webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx]: Created `SettingsGroup `

File: webapp/packages/plugin-settings-panel/src/SettingsPanelPluginBootstrap.ts (hunk 9 above)
Comment (tfokina) [reply webapp/packages/plugin-settings-panel/src/SettingsPanelPluginBootstrap.ts]: Created the action and added a handler for it

File: webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx (hunk 10 above)
Comment (tfokina) [reply webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx]: Removed this

File: webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx (hunk 11 above)
Comment (tfokina) [reply webapp/packages/plugin-settings-panel/src/SettingsPanel/SettingsPanelForm.tsx]: I agree, fixed this

**Review by Wroud — APPROVED** [review]

**Review by devnaumov — APPROVED** [review]

**Pull request merged**